scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/synchronization/Synchronization.java

```diff
@@ -0,0 +1,1130 @@
+     * <tt>Integer.MAX_VALUE</tt>.
+     *
+     * @param channel channel identifier
+     *
+     * @return the number of key-value mappings in this channel map
+     *
+     * @throws InvalidChannelException if the channel does not exist
+     */
+    int size(String channel) throws InvalidChannelException;
+
+    /**
+     * Returns <tt>true</tt> if this channel map contains no key-value mappings.
+     *
+     * @param channel channel identifier
+     *
+     * @return <tt>true</tt> if this channel map contains no key-value mappings
+     *
+     * @throws InvalidChannelException if the channel does not exist
+     */
+    boolean isEmpty(String channel) throws InvalidChannelException;
+
+    /**
+     * Returns <tt>true</tt> if this channel map contains a mapping for the specified
+     * key.  More formally, returns <tt>true</tt> if and only if
+     * this channel map contains a mapping for a key <tt>k</tt> such that
+     * <tt>(key==null ? k==null : key.equals(k))</tt>.  (There can be
+     * at most one such mapping.)
+     *
+     * @param channel channel identifier
+     * @param key key whose presence in this channel map is to be tested
+     * @return <tt>true</tt> if this channel map contains a mapping for the specified
+     *         key
+     * @throws ClassCastException if the key is of an inappropriate type for
+     *         this channel map
+     * @throws NullPointerException if the specified key is null and this channel map
+     *         does not permit null keys
+     * @throws InvalidChannelException if the channel does not exist
+     */
+    boolean containsKey(String channel, String key) throws InvalidChannelException;
+
+    /**
+     * Returns <tt>true</tt> if this channel map maps one or more keys to the
+     * specified value.  More formally, returns <tt>true</tt> if and only if
+     * this channel map contains at least one mapping to a value <tt>v</tt> such that
+     * <tt>(value==null ? v==null : value.equals(v))</tt>.  This operation
+     * will probably require time linear in the channel map size for most
+     * implementations of the <tt>Map</tt> interface.
+     *
+     * @param channel channel identifier
+     * @param value value whose presence in this channel map is to be tested
+     * @return <tt>true</tt> if this channel map maps one or more keys to the
+     *         specified value
+     * @throws ClassCastException if the value is of an inappropriate type for
+     *         this channel map
+     * (<a href="{@docRoot}/java/util/Collection.html#optional-restrictions">optional</a>)
+     * @throws NullPointerException if the specified value is null and this
+     *         channel map does not permit null values
+     * (<a href="{@docRoot}/java/util/Collection.html#optional-restrictions">optional</a>)
+     * @throws InvalidChannelException if the channel does not exist
+     */
+    boolean containsValue(String channel, Serializable value) throws InvalidChannelException;
```

Comment:
A good point, but a complex design question.

ao.getChannel() -> Channel (returns a serialized Channel object, which can only be a "proxy" to the same active object, and which contains an instance variable containing the channel name)
Problem: channel.whateverMethod(params) calls ao.whateverMethod(channel, params).
Thus, whateverMethod is also present in the ao API.

I suppose it could be handled through a good separation of interfaces, returning a proxy object which contains a stub which implements only one part of the API, etc.

Worth trying but complex

Reply:
I agree with both of your comments. Getting a Channel object when first created, and then talking to the Channel object only. That would add a layer that is easier to understand by the user I believe. Plus, we would remove one argument to basically all the methods, which is worth for the code reader.

- What would it bring to separate the interfaces? Can't the Channel and the AO both implement the same interface (why would it be bad)?

- If we do that, we could make Synchronization extends the Map interface, couldn't we ?

Reply:
I had a second thought to it. 
Strictly from the user point of view, yes it would remove the extra argument in all methods, BUT, and that's a big but, for each script which interacts with the api, it will be needed to do channel = synchronizationapi.getChannel(); channel.put("toto", "whatever").

In all examples I have done so far, the interaction with the API is mostly atomic, basically, we do one single operation per script.
This change would remove one parameter, but add another line or another function call.

I don't think it is really worth it, and especially, the complexity to do it is high.

Even with this it would not be possible to extend the Map interface, as lambdas are passed as strings.

Reply:
So, what you mean is that since channels are used in a distributed manner, getting back a Channel object when creating it would be useless since it is going to be most used at different places.

Then I agree that it is preferable to directly target the wanted action passing the channel identifier.

Your second point is indeed a blocker to extend the Map interface.